src/permutations_const.rs

```diff
@@ -0,0 +1,188 @@
+    },
+    /// No permutation left to generate.
+    End,
+}
+
+impl<I, const K: usize> fmt::Debug for PermutationsConst<I, K>
+where
+    I: Iterator + fmt::Debug,
+    I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Permutations, vals, state);
+}
+
+pub fn permutations_const<I: Iterator, const K: usize>(iter: I) -> PermutationsConst<I, K> {
+    PermutationsConst {
+        vals: LazyBuffer::new(iter),
+        state: PermutationState::Start,
+    }
+}
+
+impl<I, const K: usize> Iterator for PermutationsConst<I, K>
+where
+    I: Iterator,
+    I::Item: Clone + Default,
+{
+    type Item = [I::Item; K];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self { vals, state } = self;
+        match state {
+            PermutationState::Start => {
+                *state = PermutationState::End;
+                // TODO: Consider this case and handle it somehow, currently just using default
+                Some(std::array::from_fn(|_|I::Item::default()))
+            }
+            &mut PermutationState::Start => {
+                vals.prefill(K);
+                if vals.len() != K {
+                    *state = PermutationState::End;
+                    return None;
+                }
+                *state = PermutationState::Buffered { min_n: K };
+                let mut iter = vals[0..K].into_iter().cloned();
+                Some(std::array::from_fn(|_|iter.next().unwrap())) // TODO: Handle error case, maybe make this better
+            }
+            PermutationState::Buffered { min_n } => {
+                if vals.get_next() {
+                    let mut item = (0..K - 1)
+                        .chain(once(*min_n))
+                        .map(|i| vals[i].clone());
+                    *min_n += 1;
+                    Some(std::array::from_fn(|_|item.next().unwrap()))
+                } else {
+                    let n = *min_n;
+                    let prev_iteration_count = n - K + 1;
+                    let mut indices: Box<[_]> = (0..n).collect();
+                    let mut cycles: Box<[_]> = (n - K..n).rev().collect();
+                    // Advance the state to the correct point.
+                    for _ in 0..prev_iteration_count {
+                        if advance(&mut indices, &mut cycles) {
+                            *state = PermutationState::End;
+                            return None;
+                        }
+                    }
+                    let item = vals.get_at(&indices[0..K]); // TODO: Impl const sized variant otherwise this is pointless
+                    *state = PermutationState::Loaded { indices, cycles };
+                    Some(item.try_into().ok()?) // TODO: Handle error case
+                }
+            }
+            PermutationState::Loaded { indices, cycles } => {
+                if advance(indices, cycles) {
+                    *state = PermutationState::End;
+                    return None;
+                }
+                let k = cycles.len();
```

Comment:
Indexing into LazyBuffer would require a const-generic function also, otherwise this effort would be pointless